ibis/formats/parser.py

```diff
@@ -0,0 +1,213 @@
+import ibis.expr.datatypes as dt
+from ibis.common.bases import Singleton, Slotted
+from ibis.common.collections import FrozenDict
+
+SQLGLOT_TYPE_TO_IBIS_TYPE = {
+    DataType.Type.BIGDECIMAL: dt.Decimal(76, 38),
+    DataType.Type.BIGINT: dt.int64,
+    DataType.Type.BINARY: dt.binary,
+    DataType.Type.BIT: dt.string,
+    DataType.Type.BOOLEAN: dt.boolean,
+    DataType.Type.CHAR: dt.string,
+    DataType.Type.DATE: dt.date,
+    DataType.Type.DOUBLE: dt.float64,
+    DataType.Type.ENUM: dt.string,
+    DataType.Type.ENUM8: dt.string,
+    DataType.Type.ENUM16: dt.string,
+    DataType.Type.FLOAT: dt.float32,
+    DataType.Type.FIXEDSTRING: dt.string,
+    DataType.Type.GEOMETRY: dt.geometry,
+    DataType.Type.GEOGRAPHY: dt.geography,
+    DataType.Type.HSTORE: dt.Map(dt.string, dt.string),
+    DataType.Type.INET: dt.inet,
+    DataType.Type.INT128: dt.Decimal(38, 0),
+    DataType.Type.INT256: dt.Decimal(76, 0),
+    DataType.Type.INT: dt.int32,
+    DataType.Type.IPADDRESS: dt.inet,
+    DataType.Type.JSON: dt.json,
+    DataType.Type.JSONB: dt.json,
+    DataType.Type.LONGBLOB: dt.binary,
+    DataType.Type.LONGTEXT: dt.string,
+    DataType.Type.MEDIUMBLOB: dt.binary,
+    DataType.Type.MEDIUMTEXT: dt.string,
+    DataType.Type.MONEY: dt.int64,
+    DataType.Type.NCHAR: dt.string,
+    DataType.Type.NULL: dt.null,
+    DataType.Type.NVARCHAR: dt.string,
+    DataType.Type.OBJECT: dt.Map(dt.string, dt.json),
+    DataType.Type.SMALLINT: dt.int16,
+    DataType.Type.SMALLMONEY: dt.int32,
+    DataType.Type.TEXT: dt.string,
+    DataType.Type.TIME: dt.time,
+    DataType.Type.TIMETZ: dt.time,
+    DataType.Type.TINYINT: dt.int8,
+    DataType.Type.UBIGINT: dt.uint64,
+    DataType.Type.UINT: dt.uint32,
+    DataType.Type.USMALLINT: dt.uint16,
+    DataType.Type.UTINYINT: dt.uint8,
+    DataType.Type.UUID: dt.uuid,
+    DataType.Type.VARBINARY: dt.binary,
+    DataType.Type.VARCHAR: dt.string,
+    DataType.Type.VARIANT: dt.json,
+    DataType.Type.UNIQUEIDENTIFIER: dt.uuid,
+    #############################
+    # Unsupported sqlglot types #
+    #############################
+    # BIGSERIAL = auto()
+    # DATETIME64 = auto() # clickhouse
+    # ENUM = auto()
+    # INT4RANGE = auto()
+    # INT4MULTIRANGE = auto()
+    # INT8RANGE = auto()
+    # INT8MULTIRANGE = auto()
+    # NUMRANGE = auto()
+    # NUMMULTIRANGE = auto()
+    # TSRANGE = auto()
+    # TSMULTIRANGE = auto()
+    # TSTZRANGE = auto()
+    # TSTZMULTIRANGE = auto()
+    # DATERANGE = auto()
+    # DATEMULTIRANGE = auto()
+    # HLLSKETCH = auto()
+    # IMAGE = auto()
+    # IPPREFIX = auto()
+    # ROWVERSION = auto()
+    # SERIAL = auto()
+    # SET = auto()
+    # SMALLSERIAL = auto()
+    # SUPER = auto()
+    # TIMESTAMPLTZ = auto()
+    # UNKNOWN = auto()  # Sentinel value, useful for type annotation
+    # UINT128 = auto()
+    # UINT256 = auto()
+    # USERDEFINED = "USER-DEFINED"
+    # XML = auto()
+}
+
+
+class TypeParser(Singleton, Slotted):
```

Comment:
Prefer to use `Annotable` or rather `Concrete` (if the object is supposed to be immutable) over `Slotted` since those are way more robust. `Slotted` is mostly for reducing the boilerplate for the patterns which are dependencies for `Annotable`.

Reply:
No longer applies; I moved everything to be simple classes with only classmethods and class members.